[API] [AD] Revamp `@differentiable` usages in stdlib.

dan-zheng opened the pull request:
- Use `FloatingPoint` rather than `BinaryFloatingPoint` to constrain
  differentiability.
  - Follows from:
    - https://github.com/apple/swift/pull/21673
    - https://github.com/tensorflow/swift-bindings/pull/11
- Use `@differentiable` where clauses to constrain differentiability
  of numeric operations.
  - The most common constraint is `where Scalar : FloatingPoint` because
    `Tensor` conditionally conforms to `Differentiable where Scalar : FloatingPoint`.
- `Tensor` now conditionally conforms to `Differentiable` where `Scalar : Differentiable & FloatingPoint`.
- Allow `@differentiable` where clause conformance requirements to protocol composition types.
- Make VJP applications use the correct substitution map.
  - If a custom `@differentiable` attribute defines a VJP and where clause
    requirements, VJP applications should use a substitution map involving
    those requirements.
  - Note: more related cases need to be handled, such as `@differentiable`
    attributes with where clause requirements but no VJP. These cases will
    be handled later.

Todos:
- Make more `Tensor` operations differentiable.
  - This includes reduction and broadcasting ops.
  - This is enabled by `@differentiable` where clause type-checking.
- Use VJP functions instead of adjoint functions.
  - I've started work on this.
  - I would prefer that this be done in a separate patch, after this patch
    adds the correct `@differentiable` where clauses.
- Add tests for newly `@differentiable` `Tensor` operations.

dan-zheng committed:
[AutoDiff] [API] Revamp `@differentiable` usages in stdlib.

- Use `FloatingPoint` rather than `BinaryFloatingPoint` to constrain
  differentiability.
  - Follows from:
    - https://github.com/apple/swift/pull/21673
    - https://github.com/tensorflow/swift-bindings/pull/11
- Use `@differentiable` where clauses to constrain differentiability
  of numeric operations.
  - The most common constraint is `where Scalar : FloatingPoint` because
    `Tensor` conditionally conforms to `Differentiable where Scalar : FloatingPoint`.

Todos:
- Make more `Tensor` operations differentiable.
  - This includes reduction and broadcasting ops.
  - This is enabled by `@differentiable` where clause type-checking.
- Use VJP functions instead of adjoint functions.
  - I would prefer that this be done in a separate patch, after this patch
    adds the correct `@differentiable` where clauses.
- Add tests for newly `@differentiable` `Tensor` operations.

dan-zheng commented:
WIP, triggering tests to see what breaks.
@swift-ci Please test tensorflow

rxwei commented:
I think we should change `Tensor`'s `Differentiable` conformance condition first. Currently it is
```swift
extension Tensor : Differentiable where Scalar : FloatingPoint
```

This kinda works, but what if, in a context where `Scalar : FloatingPoint`, you want to differentiate through some scalar operation on an element of `Tensor<Scalar>`? In that case, you'll want the scalar to be differentiable as well. Therefore, we should change the conformance condition to:
```swift
extension Tensor : Differentiable where Scalar : Differentiable & FloatingPoint
```

dan-zheng committed:
[AutoDiff] Make VJP applications use the correct substitution map.

If a custom `@differentiable` attribute defines a VJP and where clause
requirements, VJP applications should use a substitution map involving
those requirements.

Note: more related cases need to be handled, such as `@differentiable`
attributes with where clause requirements but no VJP. These cases will
be handled later.

dan-zheng commented:
> Therefore, we should change the conformance condition to:
> 
> ```swift
> extension Tensor : Differentiable where Scalar : Differentiable & FloatingPoint
> ```

This makes sense. I'll adopt this change in this PR.

dan-zheng commented:
Still WIP, triggering tests.
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng committed:
[AutoDiff] [API] Change `Tensor` conditional conformance to `Differentiable`.

`Tensor` now conditionally conforms to `Differentiable` where
`Scalar : Differentiable & FloatingPoint`.

All `@differentiable` where clauses and adjoint definitions have been updated
accordingly.

Allow `@differentiable` where clause conformance requirements to protocol
composition types.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow macOS

dan-zheng commented:
@swift-ci Please test tensorflow macOS

dan-zheng commented:
@swift-ci Please test tensorflow macOS

dan-zheng commented:
@swift-ci Please test tensorflow macOS

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
Merging to unblock progress.

rxwei commented:
Would you mind changing the PR description to mention the substitution map change?

dan-zheng commented:
> Would you mind changing the PR description to mention the substitution map change?

Done! Also added info about the last commit (`Tensor : Differentiable` conditional conformance change).

rxwei commented:
> Note: more related cases need to be handled, such as `@differentiable` attributes with where clause requirements but no VJP. These cases will be handled later.

Yeah, this will be handled (and required) in differentiation through generics.
